### ISSUE TITLE
Tracker for TensorFlow-DirectML in TF 2.9.0

### ISSUE THREAD

**Issue by penpornk** [tensorflow/tensorflow#55226]
cc: @PatriceVignola @wchao1115

This issue tracks pending PRs, issues, and possible cherry-picks necessary for TensorFlow-DirectML for TF 2.9.0 release. Please post a comment with new things to track and I will update this post to reflect the changes.

PRs:
* [#54468: Enable TensorList kernels on Windows](https://github.com/tensorflow/tensorflow/pull/54468). [Approved]

**Comment by PatriceVignola** [comment]
@penpornk Another PR to add to the list: https://github.com/tensorflow/tensorflow/pull/55362

**Comment by PatriceVignola** [comment]
@penpornk How can pluggable devices implement the `Assign` op which uses the legacy variables? `TF_AssignVariable` already exists to implement `AssignVariableOp` which uses resources, but there doesn't seem to be an equivalent for the legacy variables. Should we add a `TF_AssignRefVariable` function?

**Comment by penpornk** [comment]
@PatriceVignola 

> @penpornk How can pluggable devices implement the `Assign` op which uses the legacy variables? `TF_AssignVariable` already exists to implement `AssignVariableOp` which uses resources, but there doesn't seem to be an equivalent for the legacy variables. Should we add a `TF_AssignRefVariable` function?

Looping in @wangpengmit for more info on variables.

**Comment by wangpengmit** [comment]
Resource and ref vars are very different, so my first thought is that `TF_AssignRefVariable` is probably needed.

**Comment by PatriceVignola** [comment]
I have a working prototype in a fork that I put in the kernels_experimental header. I can submit a PR if that's ok with everyone.

Also, my understanding is that ref vars are deprecated in TF2 and are replaced by resources at the python API level. Is that correct? But even though they are deprecated, some popular benchmarks like AI-Benchmark use a frozen TF1 model when running on TF2, which yield bad results for pluggable devices since they don't currently support them.

**Comment by wangpengmit** [comment]
> my understanding is that ref vars are deprecated in TF2 and are replaced by resources at the python API level. Is that correct? But even though they are deprecated, some popular benchmarks like AI-Benchmark use a frozen TF1 model when running on TF2.

Yes, ref vars are deprecated at the Python level. Existing TF1 models may still be using them, so TF2 internals still support them.

**Comment by penpornk** [comment]
> I have a working prototype in a fork that I put in the kernels_experimental header. I can submit a PR if that's ok with everyone.

@PatriceVignola Oh, that's great! Please submit the PR and we can continue the discussion there (e.g., whether this needs to be an RFC, etc). Thank you very much! :)

**Comment by PatriceVignola** [comment]
@penpork I submitted a PR here: https://github.com/tensorflow/tensorflow/pull/55379

**Comment by PatriceVignola** [comment]
New PR to add to the list: https://github.com/tensorflow/tensorflow/pull/55381

**Comment by PatriceVignola** [comment]
3 new PRs:

https://github.com/tensorflow/tensorflow/pull/55392
https://github.com/tensorflow/tensorflow/pull/55393
https://github.com/tensorflow/tensorflow/pull/55395

**Comment by PatriceVignola** [comment]
@penpornk A simple PR that was opened a month ago. I think it flew under the radar.

https://github.com/tensorflow/tensorflow/pull/54746

**Comment by PatriceVignola** [comment]
New PR: https://github.com/tensorflow/tensorflow/pull/55544

**Comment by PatriceVignola** [comment]
New PR: https://github.com/tensorflow/tensorflow/pull/55557

**Comment by PatriceVignola** [comment]
New PR: https://github.com/tensorflow/tensorflow/pull/55558

**Comment by PatriceVignola** [comment]
New PR: https://github.com/tensorflow/tensorflow/pull/55579

**Comment by PatriceVignola** [comment]
@penpornk We noticed that support for `TF_VARIANT` is not really available for pluggable devices through the API since it's very hard (or even impossible) to read the content of a C++ object through ABIs. Even if we were to use the exact same headers as TensorFlow uses for all C++ objects, we would most likely need to use the exact same compiler.

Since some models depend heavily on variant tensors which contain TensorList objects, we'd like to propose 2 new APIs that address this issue:

```cpp
TF_CAPI_EXPORT extern void TF_AddNVariant(
    TF_OpKernelContext* ctx,
    void (*binaryAddFunc)(TF_OpKernelContext* ctx, const TF_Tensor* a, const TF_Tensor* b, TF_Tensor* out),
    TF_Status* status);

TF_CAPI_EXPORT extern void TF_ZerosLikeVariant(
    TF_OpKernelContext* ctx,
    void (*zerosLikeFunc)(TF_OpKernelContext* ctx, const TF_Tensor* input, TF_Tensor* out),
    TF_Status* status);
```

Like the `TF_AssignVariable` and `TF_AssignUpdateVariable` functions currently available in `kernels_experimental.h`, the goal of `TF_AddNVariant` and `TF_ZerosLikeVariant` would be to allow plugins to implement those 2 key operations by treating the Variant objects as a black box. The Variant objects would be unwrapped within TensorFlow core, which would then call the `binaryAddFunc` or `zerosLikeFunc` functions provided by the user, which are guaranteed to contain tensors of primitives (e.g. `TF_FLOAT`).

Is this something that the TensorFlow team would like to see an RFC or PR for? For one of the RNN models that we track (Pixel-RNN), we see up to a 5x performance improvement by not having to do those operations on the CPU.

**Comment by penpornk** [comment]
> @penpornk We noticed that support for `TF_VARIANT` is not really available for pluggable devices through the API since it's very hard (or even impossible) to read the content of a C++ object through ABIs. Even if we were to use the exact same headers as TensorFlow uses for all C++ objects, we would most likely need to use the exact same compiler.

Yes, we currently don't have a generic way to support `TF_VARIANT`/`DT_VARIANT`. Even if you use the exact same header, compiler, and compiler flags, it's possible something else could still go wrong. The support so far has been on an op-by-op basis, e.g., [Kernel C API extension for Variable ops](https://github.com/tensorflow/community/pull/387), [TensorList support](https://github.com/tensorflow/tensorflow/pull/51822), and your recent PRs.

> Since some models depend heavily on variant tensors which contain TensorList objects, we'd like to propose 2 new APIs that address this issue:
> 
> ```c++
> TF_CAPI_EXPORT extern void TF_AddNVariant(
>     TF_OpKernelContext* ctx,
>     void (*binaryAddFunc)(TF_OpKernelContext* ctx, const TF_Tensor* a, const TF_Tensor* b, TF_Tensor* out),
>     TF_Status* status);
> 
> TF_CAPI_EXPORT extern void TF_ZerosLikeVariant(
>     TF_OpKernelContext* ctx,
>     void (*zerosLikeFunc)(TF_OpKernelContext* ctx, const TF_Tensor* input, TF_Tensor* out),
>     TF_Status* status);
> ```
> 
> Like the `TF_AssignVariable` and `TF_AssignUpdateVariable` functions currently available in `kernels_experimental.h`, the goal of `TF_AddNVariant` and `TF_ZerosLikeVariant` would be to allow plugins to implement those 2 key operations by treating the Variant objects as a black box. The Variant objects would be unwrapped within TensorFlow core, which would then call the `binaryAddFunc` or `zerosLikeFunc` functions provided by the user, which are guaranteed to contain tensors of primitives (e.g. `TF_FLOAT`).
> 
> Is this something that the TensorFlow team would like to see an RFC or PR for? For one of the RNN models that we track (Pixel-RNN), we see up to a 5x performance improvement by not having to do those operations on the CPU.

Thank you for the suggestion! The team thinks this sounds reasonable. If you already have a prototype code for this, would you mind opening a PR? We can take it from there. (If there are some points that need more discussion, we can start an RFC.)

**Comment by PatriceVignola** [comment]
Sure! I created a PR here: https://github.com/tensorflow/tensorflow/pull/55645

**Comment by PatriceVignola** [comment]
2 new PRs:

https://github.com/tensorflow/tensorflow/pull/55677
https://github.com/tensorflow/tensorflow/pull/55678

**Comment by PatriceVignola** [comment]
@penpornk I have a new PR to track: https://github.com/tensorflow/tensorflow/pull/56707

Do you want to create a new tracker for TF 2.10, or is this one fine?

**Comment by penpornk** [comment]
@PatriceVignola Thank you! Added to the list. 

> Do you want to create a new tracker for TF 2.10, or is this one fine?

Let's just reuse this one. :)

**Comment by PatriceVignola** [comment]
@penpornk Thank you for monitoring the other PR! Another PR that is important for us and we believe is a pretty significant bug in the pluggable device implementation is this one: https://github.com/tensorflow/tensorflow/pull/56707

The Pluggable Device RFC says that pluggable devices should be able to name themselves "GPU" and overwrite the built-in CUDA GPU, but in practice what happens is a lot of duplicate registration errors are being thrown because the previous registrations for the CUDA GPU device are not removed. This forces users to use the `tensorflow-cpu` package instead of the `tensorflow` package that GPU users are more familiar with.

**Comment by penpornk** [comment]
@PatriceVignola Happy to report that #55558 went into TF 2.10.

> Another PR that is important for us and we believe is a pretty significant bug in the pluggable device implementation is this one: https://github.com/tensorflow/tensorflow/pull/56707

Unfortunately, we need more time to carefully consider possible side effects of this one. I have replied on the PR.

I would also like to introduce @rishikasinha-tf. In the future, please help cc both her and me on PRs that got stuck. I'll also try to go through PRs that are tracked in the top post soon. Apologies again for the delay! :(

**Comment by penpornk** [comment]
@NeilGirdhar Yes, #54330 is in v2.10. We have just cut the branch on Wednesday so anything that was merged before that is in the release.

**Comment by PatriceVignola** [comment]
@penpornk We have 2 new small PRs that we'd like to get into 2.11:

https://github.com/tensorflow/tensorflow/pull/58097
https://github.com/tensorflow/tensorflow/pull/58207

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by penpornk** [comment]
Hi @Venkat6871,

Thank you for the notes. This issue is done. I'll close it.
